gpt_neox/data_utils.py

```diff
@@ -42,9 +55,151 @@ def get_tokenizer(tokenizer_type=None, from_pretrained=True, add_padding_token=F
+
+        for line_idx,path_shards,worker_shards,file_names in returns:
+            total_lines += line_idx
+            total_shards += path_shards
+            
+            max_ws = len(summary_dict['file_names'])
+            for ws in worker_shards:
+                ws[0]+=max_ws
+
+            shards.extend(worker_shards)
+            summary_dict['file_names'].extend(file_names)
+                
+    summary_dict['total_lines'] = total_lines
+    summary_dict['total_shards'] = total_shards
+
+    with jsonlines.open(output_dir+"/"+output_path, mode='w') as writer:
+        writer.write(summary_dict)
+        for shard in shards:
+            writer.write(shard)
+
+def shardify_process(worker_id,path,num_lines,chunksize,output_dir,tokenizer,seq_length):
+    start_line = worker_id*num_lines
+    end_line = (worker_id+1)*num_lines
+
+    ext_index = path.find(".")
+    last_slash = path.rfind('/')
+    path_shards,single_file_chunk,single_file_chunk_line = 0,1,0
+        
+    dataset_name = path[last_slash+1:ext_index]
+    extension = path[ext_index:]
+
+    file_names = []
+    shards = []
+
+    chunk_path=output_dir+"/"+dataset_name+"_"+str(worker_id)+"_"+str(0)+extension
+    chunk_writer = None
+
+    with jsonlines.open(path) as reader:
+        total_parsed = 0
+        for line_idx,line_loaded in enumerate(reader):
+    
+            #if total_parsed > 1000:
+            #    break 
+            if line_idx<start_line or line_idx>end_line:
+                continue
+            
+            text = line_loaded['text']
+            if len(text) == 0:
+                continue
+            total_parsed += 1
+
+            if single_file_chunk_line == 0:
+                file_names.append(chunk_path)
+                if chunk_writer:
+                    chunk_writer.close()
+                chunk_writer = jsonlines.Writer(open(chunk_path,"w"))
+ 
+            if tokenizer:
+                all_words = tokenizer(text, max_length=seq_length, return_tensors='pt',\
+                    truncation=True)['input_ids']
+                all_words = all_words.numpy().tolist()[0]
+                all_words = list(map(lambda x:int(x), all_words))
+                
+                total_words = len(all_words)
+                
+            else:
+                all_words = text.split(" ")
+                total_words = len(all_words)
+
+            #individual shards is limited by context
+            line_shards=math.floor(total_words / seq_length) + 1 
+ 
+            path_shards += line_shards
+            for i in range(line_shards):
+                #each shard contains 
+                # 1.what file we save it to (index to file_names array in summary)
+                # 2.what line of the file we're using
+                new_shard = [single_file_chunk,single_file_chunk_line+1]
+
+                start = i*seq_length
+                stop = (i+1)*seq_length 
+                trunc_words = all_words[start:stop]
```

Comment:
:thinking: I'd say it would be safer to shard by file size or something, rather than word count (1 word ~2.5 tokens so you have very different shard lengths here and it already makes the tests inconclusive)

Reply:
I would think that num tokens is the desired measurement here. Would you prefer file size to that?